Add token trait and impl for assets pallet

drewstone opened the pull request:
This PR adds a `Token` trait and implements it for `pallet_assets`. Most of the functionality is replicated with respect to the module's functional implementation with the added functionality of `transfer_from`, `allowances`, `approve`, and the addition of `Approvals` to mimic ERC20 functionality.

Related discussions:
https://github.com/paritytech/substrate/pull/7734
https://github.com/paritytech/substrate/issues/7736

Is this what you meant? @shawntabrizi

shawntabrizi commented:
@drewstone this is definitely better, but not quite there.

Let's look at just `mint` for example. There is a `mint` extrinsic and a `mint` function in your trait.

1. You want to avoid duplicate code as much as possible, so likely you just want to implement the trait, and then have the extrinsic reference your mint implementation.
2. Your mint implementation should not have authorization checks necessarily. You want like "high level functions" for other pallets. If a pallet wants to mint some token, they should just be able to do it.

Take reference from the `Currency` trait. If you call one of those functions (like `transfer`), it just happens. Authorization to call those functions happen at a higher level.


This may require a more significant amount of refactoring, but you can

shawntabrizi commented:
But yes, this is how I would approach having other pallets interact with the asset pallet.

xlc commented on frame/support/src/traits.rs:
```diff
@@ -1185,6 +1185,75 @@ pub trait VestingSchedule<AccountId> {
+	fn transfer_asset(id: Self::AssetId, origin: Source, dest: Source, amount: Self::Balance) -> DispatchResult;
+
+	/// Approve some amount of an asset to be spent by another account.
+	fn approve(id: Self::AssetId, who: Source, spender: Source, amount: Self::Balance) -> DispatchResult;
```
I really dislike the approve / transfer_from from ERC20. It makes sense at that time when ERC20 was proposed but that does mean it is still a good pattern we should follow here.

drewstone replied:
I'll add a feature flag to this as we'll likely try to mimic behavior wrt to EVM.

xlc replied:
I think another more clean way is move the erc20 feature into another trait, so the ERC20 trait does not need to be feature flag protected but the implementation can be.

shawntabrizi replied:
I agree here. don't use a feature flag.

drewstone commented:
Thanks @shawntabrizi, yea makes total sense. Will fix it up.

drewstone commented:
Addressed your comments @shawntabrizi @xlc. I've added the approvals and transfer_from behind an `erc20` feature flag and cleaned up the trait implementations.

shawntabrizi commented on frame/assets/src/lib.rs:
```diff
@@ -901,6 +786,230 @@ impl<T: Config> Module<T> {
+		})
+	}
+
+	#[cfg(feature = "erc20")]
```
I would NOT use feature flags within a runtime pallet. It is a very easy way to create a consensus bug where the native vs wam vs some other version of your runtime are not in sync, and most pipelines would not catch such an issue.

If you want to prevent these functionalities, probably a configuration trait flag would be better with some basic logic which make these functions noop in that case.

drewstone commented:
Ok, using a trait for the noop. Thanks for comments.

shawntabrizi commented on frame/assets/src/lib.rs:
```diff
@@ -153,6 +154,9 @@ pub trait Config: frame_system::Config {
 	type WeightInfo: WeightInfo;
+
+	/// Type parameter for ERC20 compatible
+	type IsERC20Compatible: Get<bool>;
```
```suggestion
	type AllowApprovalSpending: Get<bool>;
```

This is my suggestion on naming... not a big deal i guess, but this is more clear to me.

shawntabrizi replied:
To be honest it might be nice to actually have configuration flags on each of these kind of unique features, like Freezing assets or whatever.

I know in the ethereum space, they make it really flexible how you can customize the use of your token.

shawntabrizi commented:
This looks right to me overall. You need to make this compile and we should run benchmarks again.

One thing which is worse from this approach is that some storage items you have to "read twice". I.e. you do one read for authentication, and then another read/write for actually writing the changes about the token.

This should probably be okay since the storage items will always be the same, so second read will be in the overlay cache, but re-running benchmarks may tell us if there are any bigger issues.

drewstone commented:
Ok gated functionality added @shawntabrizi

shawntabrizi commented on frame/support/src/traits.rs:
```diff
@@ -1230,7 +1230,13 @@ pub trait Token<Source> {
+		spender: Source,
+		who: Source,
+		recipient: Source,
+		amount: Self::Balance
```
```suggestion
		amount: Self::Balance,
```

shawntabrizi commented:
/benchmark runtime pallet pallet_balances

shawntabrizi commented:
/benchmark runtime pallet pallet_balances

shawntabrizi commented:
@drewstone new weights look good, so overall good to me. probably want @gavofyork to have his say as well

shawntabrizi commented on frame/assets/src/lib.rs:
```diff
@@ -274,6 +288,13 @@ decl_error! {
+		/// An allowance for transfer from is insufficient
+		AllowanceTooLow,
+		/// Gated functionality errors
+		NoApprovalSpendingAllowed,
+		NoFreezingAllowed,
+		NoBurningAllowed,
+		NoMintingAllowed,
```
Doc comments needed here for metadata

shawntabrizi commented on frame/assets/src/lib.rs:
```diff
@@ -414,6 +435,7 @@ decl_module! {
 
 				*maybe_details = None;
 				Account::<T>::remove_prefix(&id);
+				Approvals::<T>::remove_prefix(&id);
```
This is not okay, you will probably need to redesign the system or remove it.

`remove_prefix` will be a storage operation per user approved, and thus can be incredibly large and not predictable/weighable.

drewstone replied:
Fine to put it behind the trait check? `if Self::ApprovalSpendingAllowed::get()`?

shawntabrizi replied:
Unfortunately this does not address my problem. The problem is that `remove_prefix` in not really that safe of an operation within the runtime and without restrictions.

It is probably better to use  a vector of users of bounded length than a hash map.

If you want to use a hash map, we might want to track somewhere the total number of approved users, and then have the user report to us how many approved users there are, we check it, and we use this for weight.

You should probably keep the approvals removal a completely separate operation with its own extrinsic so it can be measured on its own and stuff.

Also there should be a deposit placed and tracked for each approval placed on chain, since it takes up space. The deposit should be refunded when the approvals are cleaned up.

As you can see, this feature is not so trivial to do when done correctly, which is maybe why it was not included.

With something like Proxies, we probably dont need this feature anyway.

drewstone replied:
Will remove it.

drewstone commented:
Not sure if the errors are related to the last commit I made, but unsure what I'd fix to fix these. Lmk

shawntabrizi commented:
@drewstone just need to merge master

shawntabrizi commented:
@drewstone I fixed up the API and some other stuff.

Feedback for you:

* The more restrictive you make a `trait`, i.e. forcing every token to have mint, burn, freeze functionality, the less useful that trait is imo. Organize functionality into groups like I have here, and you will have something which is much more useful I think.

* StaticLookup is not free. You cannot just lookup, unlookup, back and forth. Also the low level APIs should not use StaticLookup at all. You will see all low level APIs now just use AccountId, and only one time in any spot do we do a lookup if needed.

* The configuration flags for things like AllowMinting should live at the extrinsic level, not at the low level APIs. Low level APIs should be able to do anything.

* Dont just copy and paste names. The API `transfer_ownership` made no sense at a low level API. It should be just `set_owner` at that point.


Anyway, this is a pretty good start. I think it might be worth it to update the roles API to be more flexible. We can still have a set_team extrinsic, but we should have a more generic role based access control layer at the trait level... but I think this would require a redesign.

xlc commented on frame/assets/src/lib.rs:
```diff
@@ -153,6 +157,15 @@ pub trait Config: frame_system::Config {
 	/// Weight information for extrinsics in this pallet.
 	type WeightInfo: WeightInfo;
+
+	/// The flag for determining asset freezing and thawing support.
```
I will prefer a feature flag enum instead of multiple booleans.

drewstone replied:
@shawntabrizi wdyt? Also these small changes, I'm mostly impartial to. If you want to fix please do so, I'm really just trying to get some extensible token functionality for other pallets. The faster we can merge this the better :).

gavofyork replied:
since this is meant to be going into the main codebase, i'd prefer to keep code quality high. a single flag enum is the "right" way of doing it; three separate booleans leads to boilerplatey clutter all over the codebase.

drewstone replied:
Should the enum support all combinations? W/o all combinations, something like?
```
enum AllowableFunctionalityFlag {
    Freezable,
    Mintable,
    Burnable,
    All,
    None,
}
```

xlc commented on frame/support/src/traits.rs:
```diff
@@ -1185,6 +1185,79 @@ pub trait VestingSchedule<AccountId> {
+	/// balance falls below the value of `ExistentialDeposit`, then the 'current account' is
+	/// deleted: specifically `FreeBalance`.
+	///
+	/// `system::AccountNonce` is also deleted if `ReservedBalance` is also zero (it also gets
```
this doc makes no sense

xlc commented on frame/support/src/traits.rs:
```diff
@@ -1185,6 +1185,79 @@ pub trait VestingSchedule<AccountId> {
+pub trait MintableAsset<AccountId>: FungibleAsset<AccountId> {
+	/// Increase the total issuance of of a specific asset by `amount` for a specific account.
+	///
+	/// Returns `Ok` iff the mint was successful.
```
this feels redundant

drewstone replied:
I think it’s useful for readers to have docs of return values.

xlc commented on frame/assets/src/lib.rs:
```diff
@@ -468,26 +487,14 @@ decl_module! {
 
-			Asset::<T>::try_mutate(id, |maybe_details| {
-				let details = maybe_details.as_mut().ok_or(Error::<T>::Unknown)?;
+			ensure!(T::AllowMinting::get(), Error::<T>::NoMintingAllowed);
```
not sure if we want to do feature flag check here or inside `<Self as MintableAsset<_>>::mint`

drewstone replied:
This is what @shawntabrizi said we should not do.

shawntabrizi replied:
You can see that the low level APIs do not have any of these feature flag stuff. You can import this module without any `Call` and use it simple as the backend for any token system you desire. As such, your interaction layer should be the one to check if minting is allowed, or whether you even want to constrain to this trait when using this Asset.

But would be interesting to see if you disagree here.

xlc replied:
I guess the ambiguity lies in the purpose the flag, does it protect only the extrinsics or protect the whole pallet.
Improved docs should make this clear it only protect the extrinsics.

shawntabrizi replied:
Sure, documentation will help here.

But here is an example why the flag i think could never apply at the whole pallet. Imagine some other pallet which needs an implementer of `FungibleAsset + MintableAsset + BurnableAsset`. Let's say the functionality of this pallet depends on these traits. If we then allowed the pallet to disable any of these functions at the pallet level, Rust will allow this pallet to satisfy those trait conditions, but we would know that it wouldn't work.

I guess it would be that for this kind of scenario, feature flags could allow us to include certain trait implementations dynamically, but as I mentioned, feature flags are a little bit scary to me for the runtime.

xlc replied:
Cargo feature will make sure you get compile time error. But it have its own problem.

drewstone commented:
How can I help push this through? Or are these comments above unresolved @shawntabrizi @xlc @gavofyork ?

gavofyork commented on frame/assets/src/lib.rs:
```diff
@@ -901,6 +800,191 @@ impl<T: Config> Module<T> {
+		who: T::AccountId,
+		amount: Self::Balance
+	) -> DispatchResult {
+		Asset::<T>::try_mutate(id, |maybe_details| {
```
if it's the pallet which (through type `T`) can be configured to be non-burnable, then why would this implementation (over `T`) not adhere to that?

drewstone replied:
Not sure I understand entirely, are you saying its best to move these implementations into other repos that want this functionality?

gavofyork commented on bin/node/runtime/src/lib.rs:
```diff
@@ -940,6 +940,9 @@ impl pallet_mmr::Config for Runtime {
 parameter_types! {
 	pub const AssetDepositBase: Balance = 100 * DOLLARS;
 	pub const AssetDepositPerZombie: Balance = 1 * DOLLARS;
+	pub const AllowFreezing: bool = true;
```
what's the point of adding these? is there some good reason by a chain might want to limit the functionality over all assets it hosts?

drewstone replied:
This was requested in eariler reviews afaik. At first all this functionality was supported by we couldn't come to consensus.

sinkingsugar commented:
Why this PR never made it?
pallet-assets is 100% useless without...

shawntabrizi commented:
@sinkingsugar there are traits which give access to all the token features in assets: 
https://crates.parity.io/frame_support/traits/tokens/fungible/index.html

sinkingsugar commented:
> @sinkingsugar there are traits which give access to all the token features in assets: https://crates.parity.io/frame_support/traits/tokens/fungible/index.html

Thank you @shawntabrizi - I already forked substrate to make it work but will gladly revert!
This could use some love especially @ pallet-assets readme level (it's extremely outdated)
